coordinator/src/main/scala/filodb.coordinator/FilodbCluster.scala

```diff
@@ -172,50 +174,45 @@ final class FilodbCluster(val system: ExtendedActorSystem) extends Extension wit
 
-  /** Begins node graceful shutdown:
-    *   A. Sets `isTerminating` to true and `isJoined` and `isInitialized` to false,
-    *   B. Calls `akka.cluster.Cluster.leave(selfAddress)`
-    *   C. Awaits the async graceful shutdown of the `filodb.coordinator.NodeGuardian`
-    *      supervised child actors via `filodb.coordinator.GracefulStopStrategy`
-    *   D. Shuts down `monix.execution.Scheduler.io` and `filodb.core.store.StoreFactory`'s MetaStore and MemStore
-    *   E. Calls the async terminate on `akka.actor.ActorSystem`
-    *
-    * Idempotent.
-    */
-  protected[filodb] def shutdown(): Unit =
-    if (!isTerminated) {
-      import akka.pattern.gracefulStop
-
-      import NodeProtocol.GracefulShutdown
-
-      logger.info("Leaving the cluster.")
-      _cluster.get foreach (_.leave(selfAddress))
+  /**
+   * Hook into Akka's CoordinatedShutdown sequence
+   * Please see https://doc.akka.io/docs/akka/current/actors.html#coordinated-shutdown for details
+   */
+  CoordinatedShutdown(system).addTask(CoordinatedShutdown.PhaseServiceUnbind, "queryShutdown") { () =>
+    implicit val timeout = Timeout(15.seconds)
+    // Reset shuts down all ingestion and query actors on this node
+    // TODO: be sure that status gets updated across cluster?
+    (coordinatorActor ? NodeProtocol.ResetState).map(_ ⇒ Done)
+  }
 
-      _isInitialized.set(false)
-      logger.info("Terminating: starting shutdown")
+  // TODO: hook into service-stop "forcefully kill connections?"  Maybe send each outstanding query "TooBad" etc.
 
-      try {
-        if (clusterActor.isDefined) {
-          Await.result(gracefulStop(guardian, GracefulStopTimeout, GracefulShutdown), GracefulStopTimeout)
-        }
+  CoordinatedShutdown(system).addTask(CoordinatedShutdown.PhaseBeforeClusterShutdown, "storeShutdown") { () =>
+    _isInitialized.set(false)
+    logger.info("Terminating: starting shutdown")
 
-        system.terminate foreach { _ =>
-          logger.info("Actor system was shut down")
-        }
-        metaStore.shutdown()
-        memStore.shutdown()
+    try {
+      metaStore.shutdown()
+      memStore.shutdown()
+      ioPool.shutdown()
+    } catch {
+      case NonFatal(e) =>
+        system.terminate()
         ioPool.shutdown()
-      } catch {
-        case NonFatal(e) =>
-          system.terminate()
-          ioPool.shutdown()
-      }
-      finally _isTerminated.set(true)
     }
+    finally _isTerminated.set(true)
 
-  Runtime.getRuntime.addShutdownHook(new Thread() {
-    override def run(): Unit = shutdown()
-  })
+    Future.successful(Done)
+  }
+
+  /**
+   * Invokes CoordinatedShutdown to shut down the ActorSystem, leave the Cluster, and our hooks above
+   * which shuts down the stores and threadpools.
+   * NOTE: Depending on the setting of coordinated-shutdown.exit-jvm, this might cause the JVM to exit.
+   */
+  protected[filodb] def shutdown(): Unit = {
+    CoordinatedShutdown(system).run(CoordinatedShutdown.UnknownReason)
```

Comment:
If you haven't already, I think you can now remove [filodb.coordinator.NodeProtocol.GracefulShutdown](https://github.com/filodb/FiloDB/blob/develop/coordinator/src/main/scala/filodb.coordinator/NodeGuardian.scala#L167) and its usage in specs, and previous custom logic around `filodb.coordinator.GracefulStopAwareSupervisor.gracefulShutdown` 
- https://github.com/filodb/FiloDB/blob/develop/coordinator/src/main/scala/filodb.coordinator/BaseActor.scala#L80
- https://github.com/filodb/FiloDB/blob/develop/coordinator/src/main/scala/filodb.coordinator/BaseActor.scala#L86
- https://github.com/filodb/FiloDB/blob/develop/coordinator/src/main/scala/filodb.coordinator/BaseActor.scala#L94-L107 

I no longer see anything using [GracefulStopStrategy](https://github.com/filodb/FiloDB/blob/develop/coordinator/src/main/scala/filodb.coordinator/GracefulStopStrategy.scala#L17), and there's duplicate code, so perhaps that can be removed as well. I think originally one was for supervisors, like NodeGuardian, and one for child workers to handle graceful shutdown of their processing when possible.

Reply:
Yes, thanks